docs/weave/weave-api-spec/03-querying.md

```diff
@@ -0,0 +1,192 @@
+---
+id: weave-query-spec
+title: Querying Weave
+sidebar_label: Querying Weave
+---
+
+In this section querying Weave for data will be explained. Before diving into the details, Weave internals such as __buckets__ and __paths__ must be explained.
+
+## Buckets
+
+Buckets are the structures that enable accessing and writing to __Key-Value__ database in an organized and controlled manner. Buckets stores the data as well as the indexes, secondary indexes that refer to the data.
+
+### Accessing buckets
+
+[//]: # (TODO give reference to Weave/tendermint or ABCI documentation)
+As mentioned in the previous sections, Weave uses tendermint as consensus engine thus queries are made to data store via `abci_queries`. Therefore when you make a query actually you do the call to tendermint's `ABCI` protocol. For more info about underlyings refer to [tendermint/abciquery](https://tendermint.com/rpc/#abciquery).
+
+Via running the JSON-RPC/HTTP call below, __hugnet__ testnet could be queried so you can see an example response.
+
+``` bash
+curl -X POST -d '{ "json-rpc": 2.0, "id": "foobar321", "method": "abci_query", "params": { "path": "/tokens", "data": "" } }' https://bns.hugnet.iov.one/
+```
+
+### Bucket paths
+
+A bucket data can be queried using that bucket unique path to identify it.
+The above curl command is reading username Token entity using */tokens* path.
+
+Some available bucket paths: `/wallets`, `/auth`, `/aswaps` ...
+
+## Indexes
+
+### Primary indexes
+
+For accessing the data resides inside buckets, indexes are used. E.g.  to access wallet with primary index `00CAFE00`__(hex)__, call has to be made to `/wallets` path with index as data.
+
+```bash
+curl -X POST -d '{ "json-rpc": 2.0, "id": "foobar321", "method": "abci_query", "params": { "path": "/wallets", "data": "00CAFE00" } }' https://bns.antnet.iov.one/
+```
+
+- Path: ``/``, Data: ``0123456789`` (hex) -> db.Get(``0123456789``)
+  - Queries made to root (`/`) are direct queries to the key-value db without an index.
+
+- Path: ``/wallets``, Data: ``00CAFE00`` (hex) -> wallets.Get(``00CAFE00``)
+  - `wallets` are queried for the account with address `00CAFE00`.
+
+### Secondary indexes
+
+Another way to access data is using __secondary indexes__. Via secondary indexes data could have multiple ways to be accessed. E.g. wallets are registered under a name so there might be some use cases for accessing them with names. `/wallets` + `/name` = `/wallets/name` indicates that bucket will be queried using name index that will be sent in the data field.
+
+- Path: ``/wallets/name``, Data: "John" (raw): wallets.Index("name").Get("John")
+  - `wallets` are queried for the account with name `John`.
+
+### Multikey indexes
+
+There might be some cases where one index have multiple values. Multikeys exists for this purpose. Query response will be multiple values instead of one.
+
+### Prefixes
+
+There might be cases which all the data with index that begins with prefix. `wallets`
```

Comment:
Please add the useful case to "list all tokens", "list all wallets" where data=""